multi/gr716-multi/gr716-multi.c

```diff
@@ -0,0 +1,323 @@
+/*
+ * Phoenix-RTOS
+ *
+ * GR716 multi driver main
+ *
+ * Copyright 2023 Phoenix Systems
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <board_config.h>
+#include <errno.h>
+#include <paths.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <libklog.h>
+
+#include <sys/debug.h>
+#include <sys/file.h>
+#include <sys/msg.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/threads.h>
+#include <posix/utils.h>
+
+#include "adc.h"
+#include "gpio.h"
+#include "spi.h"
+#include "uart.h"
+#include "gr716-multi.h"
+
+
+#define UART_THREADS_NO  2
+#define MULTI_THREADS_NO 2
+#define GR716_MULTI_PRIO 3
+
+#define STACKSZ 2048
+
+#define LOG(fmt, ...)       printf("gr716-multi: " fmt "\n", ##__VA_ARGS__)
+#define LOG_ERROR(fmt, ...) fprintf(stderr, "gr716-multi: " fmt "\n", ##__VA_ARGS__)
+
+
+static const char *multi_devs[] = { "gpio0", "gpio1", "spi0", "spi1", "uart2", "adc0", NULL };
+
+static struct {
+	oid_t multiOid;
+	oid_t uartOid;
+	char uartStack[UART_THREADS_NO][STACKSZ] __attribute__((aligned(8)));
+	char stack[MULTI_THREADS_NO - 1][STACKSZ] __attribute__((aligned(8)));
+} multi_common;
+
+
+static int multi_createDevs(void)
+{
+	int err;
+
+	for (int i = 0; multi_devs[i] != NULL; i++) {
+		err = create_dev(&multi_common.multiOid, multi_devs[i]);
+		if (err < 0) {
+			LOG_ERROR("Failed to create %s device file", multi_devs[i]);
+			return err;
+		}
+	}
+
+	return 0;
+}
+
+
+static id_t multi_getId(msg_t *msg)
+{
+	id_t id = 0;
+
+	switch (msg->type) {
+		case mtDevCtl:
+			id = ((multi_i_t *)msg->i.raw)->id;
+			break;
+		case mtRead:
+		case mtWrite:
+			id = msg->i.io.oid.id;
+			break;
+		case mtGetAttr:
+		case mtSetAttr:
+			id = msg->i.attr.oid.id;
+			break;
+		default:
+			break;
+	}
+
+	return id;
+}
+
+
+static void multi_dispatchMsg(msg_t *msg)
+{
+	id_t id = multi_getId(msg);
+
+	switch (id) {
+		case id_gpio0:
+		case id_gpio1:
+			gpio_handleMsg(msg, id);
+			break;
+
+		case id_spi0:
+		case id_spi1:
+			spi_handleMsg(msg, id);
+			break;
+
+		case id_adc0:
+		case id_adc1:
+		case id_adc2:
+		case id_adc3:
+		case id_adc4:
+		case id_adc5:
+		case id_adc6:
+		case id_adc7:
+			adc_handleMsg(msg, id);
+			break;
+
+		default:
+			msg->o.io.err = -EINVAL;
+			break;
+	}
+}
+
+
+static void multi_thread(void *arg)
+{
+	msg_t msg;
+	unsigned long int rid;
+
+	(void)arg;
+
+	while (1) {
+		while (msgRecv(multi_common.multiOid.port, &msg, &rid) < 0) {
+		}
+
+		switch (msg.type) {
+			case mtRead:
+			case mtWrite:
+			case mtGetAttr:
+			case mtSetAttr:
+			case mtDevCtl:
+				multi_dispatchMsg(&msg);
+				break;
+
+			case mtOpen:
+			case mtClose:
+				msg.o.io.err = EOK;
+				break;
+
+			case mtCreate:
+				msg.o.create.err = -ENOSYS;
+				break;
+
+			case mtTruncate:
+			case mtDestroy:
+			case mtLookup:
+			case mtLink:
+			case mtUnlink:
```

Comment:
Better to return 0 on link and unlink - if we return error on unlink sad stuff happens when removing special file - error is returned in shell, but the file is still removed from the filesystem